File: crates/taflow-core/src/stream/normalized_average_true_range.rs
//! Batch implementation for `normalized_average_true_range`.

use super::aroon_true_range::*;
use crate::error::{TaError, TaResult};

/// Compute the normalized average true range result for the supplied aligned series.
///
/// # Parameters
///
/// * `high` - Input series or configuration value.
/// * `low` - Input series or configuration value.
/// * `close` - Input series or configuration value.
/// * `timeperiod` - Input series or configuration value.
///
/// # Returns
///
/// An aligned result with TA-Lib-compatible validation and warm-up values.
pub fn normalized_average_true_range(
    high: &[f64],
    low: &[f64],
    close: &[f64],
    timeperiod: usize,
) -> TaResult<Vec<f64>> {
    if high.len() != low.len() || high.len() != close.len() {
        return Err(crate::TaError::LengthMismatch {
            expected: high.len(),
            got: low.len().min(close.len()),
        });
    }
    let mut state = NormalizedAverageTrueRange::new(timeperiod)?;
    Ok(high
        .iter()
        .zip(low)
        .zip(close)
        .map(|((high, low), close)| state.append(*high, *low, *close).unwrap_or(f64::NAN))
        .collect())
}
use super::*;

/// Stateful normalized ATR, matching `NATR = ATR / close * 100`.
#[derive(Debug, Clone)]
/// Persistent Rust state or aligned output type for `NormalizedAverageTrueRange`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct NormalizedAverageTrueRange {
    atr: AverageTrueRange,
    value: Option<f64>,
}

impl NormalizedAverageTrueRange {
    /// Computes or updates `new` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn new(period: usize) -> TaResult<Self> {
        Ok(Self {
            atr: AverageTrueRange::new(period)?,
            value: None,
        })
    }

    /// Computes or updates `append` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn append(&mut self, high: f64, low: f64, close: f64) -> Option<f64> {
        self.value = self.atr.append(high, low, close).map(|atr| {
            if close == 0.0 {
                0.0
            } else {
                atr / close * 100.0
            }
        });
        self.value
    }

    /// Computes or updates `value` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Computes or updates `reset` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn reset(&mut self) {
        self.atr.reset();
        self.value = None;
    }
}

File: python/taflow/indicators/normalized_average_true_range.py
"""Persistent Normalized Average True Range adapter."""

from typing import Any

import numpy as np

from .._native import NormalizedAverageTrueRange as _NativeNormalizedAverageTrueRange
from .._series import as_float64_series


class NormalizedAverageTrueRange:
    """Compute ``100 * AverageTrueRange / close`` in persistent Rust state.

    Supply aligned high, low, and close histories through ``extend``. ``timeperiod`` defaults to 14 and must be positive. Warm-up is
    NaN through index ``timeperiod - 1`` and a warmed zero close returns zero.
    TA-Lib's historical period-1 contract returns raw True Range without
    normalization; this adapter preserves that special case. This maps to
    TA-Lib ``NATR``.
    """

    def __init__(self, timeperiod: int = 14) -> None:
        self._state = _NativeNormalizedAverageTrueRange(timeperiod)

    def append(
        self, high: float, low: float, close: float
    ) -> "NormalizedAverageTrueRange":
        """Append one high/low/close tuple and return this indicator."""
        self._state.append(float(high), float(low), float(close))
        return self

    def extend(self, high: Any, low: Any, close: Any) -> "NormalizedAverageTrueRange":
        """Append aligned high, low, and close histories and return this indicator."""
        self._state.extend(
            as_float64_series(high), as_float64_series(low), as_float64_series(close)
        )
        return self

    def compute(self) -> np.ndarray:
        """Return the aligned float64 history with NaN warm-up."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest value, or ``None`` during warm-up."""
        return self._state.value

    def reset(self) -> "NormalizedAverageTrueRange":
        """Restore fresh native state, clear history, and return this indicator."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of processed tuples."""
        return len(self._state)
